### ISSUE TITLE
RangeError: Maximum call stack size exceeded during refactoring

### ISSUE THREAD

**Issue by ngub** [Microsoft/TypeScript#17696]
**TypeScript Version:**  2.4.1

Currently I am in process of refactoring of project file structure.
After moving all the files to their new locations I am fixing all the broken paths using tsc output for hints.

Suddenly, after fixing a certain amount of import errors tsc fails with this error. I've tried all the process from scratch several times fixing broken imports manually or using IDE refactoring feature - the result is the same. Also I've been repeating the process in different orders, trying to move and fix each file one by one or moving all files in bulk and fix imports after moving.

Unfortunately, I cannot provide more clues but will update this issue if I find any.

```ts
RangeError: Maximum call stack size exceeded
at resolveAlias (/home/nikita/projects/balance-of-me-admin/node_modules/gulp-typescript/node_modules/typescript/lib/typescript.js:25256:22)
at checkIdentifier (/home/nikita/projects/balance-of-me-admin/node_modules/gulp-typescript/node_modules/typescript/lib/typescript.js:33458:113)
at checkExpressionWorker (/home/nikita/projects/balance-of-me-admin/node_modules/gulp-typescript/node_modules/typescript/lib/typescript.js:37914:28)
at checkExpression (/home/nikita/projects/balance-of-me-admin/node_modules/gulp-typescript/node_modules/typescript/lib/typescript.js:37894:42)
at checkExpressionWithContextualType (/home/nikita/projects/balance-of-me-admin/node_modules/gulp-typescript/node_modules/typescript/lib/typescript.js:37781:26)
at checkApplicableSignature (/home/nikita/projects/balance-of-me-admin/node_modules/gulp-typescript/node_modules/typescript/lib/typescript.js:35759:35)
at chooseOverload (/home/nikita/projects/balance-of-me-admin/node_modules/gulp-typescript/node_modules/typescript/lib/typescript.js:36238:30)
at resolveCall (/home/nikita/projects/balance-of-me-admin/node_modules/gulp-typescript/node_modules/typescript/lib/typescript.js:36146:26)
at resolveCallExpression (/home/nikita/projects/balance-of-me-admin/node_modules/gulp-typescript/node_modules/typescript/lib/typescript.js:36324:20)
at resolveSignature (/home/nikita/projects/balance-of-me-admin/node_modules/gulp-typescript/node_modules/typescript/lib/typescript.js:36515:28)
```

**Comment by ngub** [comment]
Okay, I've found the cause.

In one of the components I had the following case.
The component's index file has `import ChildComponent from './ChildComponent'`
In the component's folder there are ./ChildComponent.ts (a file) and ./ChildComponent folder with index.ts inside. Both ./ChildComponent.ts and ./ChildComponent/index.ts have default exports.

Looks like this causes tsc to enter infinite recursion. After I resolved this conflict, the error disappeared.

Although, the problem was caused by me derping, I think, this is *unexpected behavior*.
*Expected behavior*
Fail with error message like this: `Error: ambiguous import`
